### PR TITLE
Develop task forms

### PR DESCRIPTION
**Pull request by scorpionka**

### CODE REVIEW

File: src/app/shared/components/course-form/course-form.component.ts [hunk 1]
```diff
@@ -1,19 +1,61 @@
...
+  durationValue?: number = 0;
+
+  ngOnInit(): void {
+    this.courseForm = this.fb.group({
+      title: ["", [Validators.required, Validators.minLength(2)]],
+      description: ["", [Validators.required, Validators.minLength(2)]],
+      author: ["", [Validators.minLength(2)]],
+      duration: ["", [Validators.required, Validators.min(1)]],
+    });
+  }
```
Comment (UladzislauMazhanau) [inline src/app/shared/components/course-form/course-form.component.ts:30]: For authors, you don't need a separate component and a separate form according to the task conditions. You should have two FormArrays to store the list of authors and the list of course authors, as well as a '`newAuthor`' field that will be a FormGroup. So, the basic structure of the form should look something like this:
```suggestion
this.courseForm = this.fb.group({
      title: ['', [Validators.required, Validators.minLength(2)]],
      description: ['', [Validators.required, Validators.minLength(2)]],
      duration: ["", [Validators.required, Validators.min(1)]],
      authors: this.fb.array([]),
      courseAuthors: this.fb.array([]),
      newAuthor: this.fb.group({
        authorName: ['', [Validators.minLength(2), Validators.pattern('^[a-zA-Z0-9 ]*$')]]
      })
    });
```
 When we select an author from the '`authors`' list, they are removed from that list and added to the '`courseAuthors`' list. When removed from '`courseAuthors`', they return to the '`authors`' array. To work with FormArrays, you can refer to this article - [[link](https://netbasal.com/angular-reactive-forms-the-ultimate-guide-to-formarray-3adbe6b0b61a) If you are using the 'input' tag to display values from the authors' list, consider using the 'disabled' attribute like this: [disabled]="true".]. For working with a nested FormGroup ('newAuthor'), you can refer to this article - [[link](https://angular.io/api/forms/FormGroupName#register-a-nested-formgroup)].

**Review by UladzislauMazhanau — CHANGES_REQUESTED** [review]

**Comment by UladzislauMazhanau** [comment]
[Optional] It would be great if you also covered styles for fields in the 'ng-invalid' state since it's a common practice when working with forms. In your current setup, fields always start with the 'ng-invalid' state, and the user sees everything in red even before entering any data or interacting with the form, which is not ideal.

File: src/app/shared/components/course-form/course-form.component.ts [hunk 2]
```diff
@@ -1,19 +1,61 @@
...
+  get duration() {
+    return this.courseForm.get("duration");
+  }
+
+  onSubmit() {
+    if (this.courseForm.valid) {
+      console.log("The form was submitted", this.courseForm.value);
+      this.router.navigate(["/"]);
+    } else console.log("The form is invalid.", this.courseForm.value);
+  }
```
[the remark this answers is not in the document]
Comment (scorpionka) [reply src/app/shared/components/course-form/course-form.component.ts:53]: done

File: src/app/shared/components/registration-form/registration-form.component.ts [hunk 3]
```diff
@@ -1,12 +1,41 @@
...
+  get password() {
+    return this.registrationForm.get("password");
+  }
+
+  onSubmit() {
+    if (this.registrationForm.valid) {
+      console.log("The form was submitted", this.registrationForm.value);
+      this.router.navigate(["/"]);
+    } else console.log("The form is invalid.", this.registrationForm.value);
+  }
```
[the remark this answers is not in the document]
Comment (scorpionka) [reply src/app/shared/components/registration-form/registration-form.component.ts:40]: done

File: src/app/shared/components/course-form/course-form.component.html [hunk 4]
```diff
@@ -1,14 +1,28 @@
...
 <h2>Create / Edit Course</h2>
-<form>
+<form
+  class="create-edit-course-form"
+  [formGroup]="courseForm"
+  (ngSubmit)="onSubmit()"
+>
   <div class="app-plate">
-    <div class="course">
+    <div class="course" class="course-form">
```
[the remark this answers is not in the document]
Comment (scorpionka) [reply src/app/shared/components/course-form/course-form.component.html:9]: fixed

File: src/app/shared/components/course-form/course-form.component.html [hunk 5]
```diff
@@ -19,38 +33,88 @@ <h3>Main Info</h3>
...
+            Description is required.
+          </div>
+          <div *ngIf="description!.errors?.['minlength']">
+            Description should not be less than 2 symbols.
+          </div>
+        </div>
       </div>
       <div class="separator"></div>
+      <h3>Duration</h3>
+      <div class="form__control" class="half-size">
```
[the remark this answers is not in the document]
Comment (scorpionka) [reply src/app/shared/components/course-form/course-form.component.html:52]: fixed

File: src/app/shared/components/course-form/course-form.component.html [hunk 6]
```diff
@@ -19,38 +33,88 @@ <h3>Main Info</h3>
...
+            Duration is required.
+          </div>
+          <div *ngIf="duration!.errors?.['min']">
+            Duration should be greater than 0.
+          </div>
+        </div>
+      </div>
       <h3>Authors</h3>
-      <div class="form__control">
+      <div class="form__control" class="half-size">
```
[the remark this answers is not in the document]
Comment (scorpionka) [reply src/app/shared/components/course-form/course-form.component.html:73]: The implementation was changed.

File: src/app/shared/components/course-form/course-form.component.html [hunk 7]
```diff
@@ -19,38 +33,88 @@ <h3>Main Info</h3>
...
-        </div>
-        <span id="durationErrorMessage" class="text-danger">Duration is required.</span>
-      </div>
-      <div class="form__action">
-        <app-button></app-button>
-      </div>
     </div>
   </div>
 </form>
+<div class="form__action" class="course-form-button">
```
[the remark this answers is not in the document]
Comment (scorpionka) [reply src/app/shared/components/course-form/course-form.component.html:108]: fixed

File: src/app/shared/components/login-form/login-form.component.html [hunk 8]
```diff
@@ -1,24 +1,71 @@
...
-          <span id="emailErrorMessage" class="text-danger">Email is required.</span>
+          <input
+            type="text"
+            class="form-control"
+            id="email"
+            placeholder="Input text"
+            name="email"
+            [(ngModel)]="userModel.email"
+            #emailLoginControl="ngModel"
+            name="email"
```
[the remark this answers is not in the document]
Comment (scorpionka) [reply src/app/shared/components/login-form/login-form.component.html:17]: The duplicate was removed.

File: src/app/shared/components/login-form/login-form.component.html [hunk 9]
```diff
@@ -1,24 +1,71 @@
...
+            class="text-danger"
+          >
+            Password is required.
+          </span>
         </div>
       </div>
       <div>
-        <div class="form__action">
-          <app-button></app-button>
+        <div class="form__action" class="login-submit-button">
```
[the remark this answers is not in the document]
Comment (scorpionka) [reply src/app/shared/components/login-form/login-form.component.html:61]: fixed

File: src/app/shared/components/registration-form/registration-form.component.html [hunk 10]
```diff
@@ -1,28 +1,68 @@
...
+          formControlName="password"
+        />
+        <div *ngIf="password?.invalid && password!.touched" class="text-danger">
+          Password is required.
+        </div>
       </div>
     </div>
-    <div class="form__action">
-      <app-button></app-button>
+    <div class="form__action" class="registration-form-button">
```
[the remark this answers is not in the document]
Comment (scorpionka) [reply src/app/shared/components/registration-form/registration-form.component.html:56]: fixed

File: src/app/shared/components/course-form/course-form.component.html [hunk 11]
```diff
@@ -19,38 +33,88 @@ <h3>Main Info</h3>
...
+      <div class="form__control" class="half-size">
+        <div class="course__flex">
+          <!--Use the name `duration` for the form array.-->
+          <input
+            type="number"
+            id="duration"
+            placeholder="Input duration"
+            formControlName="duration"
+          />
+          <div class="course__duration">{{ durationValue! | duration }}</div>
```
[the remark this answers is not in the document]
Comment (scorpionka) [reply src/app/shared/components/course-form/course-form.component.html:61]: fixed

File: src/app/shared/components/registration-form/registration-form.component.ts [hunk 12]
```diff
@@ -1,12 +1,41 @@
...
+export class RegistrationFormComponent implements OnInit {
   registrationForm!: FormGroup;
-  // Use the names `name`, `email`, `password` for the form controls.
+
+  constructor(private formBuilder: FormBuilder, private router: Router) {}
+
+  ngOnInit(): void {
+    this.registrationForm = this.formBuilder.group({
+      name: ["", [Validators.required, Validators.minLength(6)]],
+      email: ["", [Validators.required]],
```
[the remark this answers is not in the document]
Comment (scorpionka) [reply src/app/shared/components/registration-form/registration-form.component.ts:18]: added

File: src/app/shared/components/registration-form/registration-form.component.html [hunk 13]
```diff
@@ -1,28 +1,68 @@
...
         <!--Use the name `email` for the form control.-->
         <label for="email">Email</label>
-        <input type="text" id="email" placeholder="Input text"/>
-        <span id="emailErrorMessage" class="text-danger">Email is required.</span>
+        <input
+          type="text"
+          id="email"
+          placeholder="Input text"
+          formControlName="email"
+          appEmailValidator="[a-z0-9]+@[a-z]+.[a-z]{2,3}"
```
[the remark this answers is not in the document]
Comment (scorpionka) [reply src/app/shared/components/registration-form/registration-form.component.html:33]: removed

File: src/app/shared/components/course-form/course-form.component.ts (hunk 1 above)
Comment (scorpionka) [reply src/app/shared/components/course-form/course-form.component.ts]: done

File: src/app/shared/components/login-form/login-form.component.html [hunk 14]
```diff
@@ -1,24 +1,70 @@
...
+            "
+            class="text-danger"
+          >
+            <div *ngIf="emailLoginControl.errors?.['required']">
+              Email is required.
+            </div>
+            <div *ngIf="emailLoginControl.errors?.['emailPattern']">
+              Email format is invalid.
+            </div>
+          </span>
```
Comment (UladzislauMazhanau) [inline src/app/shared/components/login-form/login-form.component.html:33]: Better not to use block-level elements like `<div>` inside an inline element like `<span>`. In your case, it's better to do the opposite and use `<span>` inside `<div>`. Also, in order not to display multiple errors at the same time, you can use `ngIf` like this : `<div *ngIf="condition; else elseBlock">Content to render when condition is true.</div>
<ng-template #elseBlock>Content to render when condition is false.</ng-template>` ([Link](https://angular.io/api/common/NgIf#description))

**Comment by UladzislauMazhanau** [comment]
Good job!

**Pull request merged**